state/upgrades.go

```diff
@@ -518,84 +518,109 @@ func addNonDetachableStorageMachineId(st *State) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	getPool := func(d bson.M) string {
+		var pool string
+		info, _ := d["info"].(bson.M)
+		params, _ := d["params"].(bson.M)
+		if info != nil {
+			pool = info["pool"].(string)
+		} else if params != nil {
+			pool = params["pool"].(string)
+		}
+		return pool
+	}
+
+	var needsUpgradeTerm = bson.D{
+		{"machineid", bson.D{{"$exists", false}}},
+		{"hostid", bson.D{{"$exists", false}}},
+	}
 	var ops []txn.Op
-	volumes, err := sb.volumes(
-		bson.D{{"machineid", bson.D{{"$exists", false}}}},
-	)
-	if err != nil {
+
+	volumeColl, cleanup := st.db().GetCollection(volumesC)
+	defer cleanup()
+
+	var volData []bson.M
+	err = volumeColl.Find(needsUpgradeTerm).All(&volData)
+	if err != nil && err != mgo.ErrNotFound {
 		return errors.Trace(err)
 	}
-	for _, v := range volumes {
-		var pool string
-		if v.doc.Info != nil {
-			pool = v.doc.Info.Pool
-		} else if v.doc.Params != nil {
-			pool = v.doc.Params.Pool
-		}
-		detachable, err := isDetachableVolumePool(sb, pool)
+
+	volumeAttachColl, cleanup := st.db().GetCollection(volumeAttachmentsC)
+	defer cleanup()
+
+	for _, v := range volData {
+		detachable, err := isDetachableVolumePool(sb, getPool(v))
 		if err != nil {
 			return errors.Trace(err)
 		}
 		if detachable {
 			continue
 		}
-		attachments, err := sb.VolumeAttachments(v.VolumeTag())
-		if err != nil {
+
+		var attachInfo []bson.M
+		err = volumeAttachColl.Find(bson.D{{"volumeid", v["name"]}}).All(&attachInfo)
```

Comment:
Are we worried about the n+1 query structure here? Would it be worth getting all the volume attachments in one query and loading them into a map?